### DISCUSSION TITLE
String matching

### DISCUSSION THREAD

**Discussion by rolyhudson** [BHoM/BHoM#1156] [Ideas]
Setting this up to keep the discussion from the string matching workshop live and open for all.

Notes from the workshop are [here](https://burohappold.sharepoint.com/:f:/r/sites/BHoM/01_Master%20File/07_Meetings/210107_String%20Matching%20Workshop?csf=1&web=1&e=eLqTgx).
Participants in that session @alelom, @AndreBalasian, @adecler, @FraserGreenroyd, @LMarkowski, @pawelbaran, @rolyhudson, + others @al-fisher, @IsakNaslundBh 

Follow-on conversation with @pawelbaran deemed it a good idea to set up an open discussion.

Good starting point as a reference / source for string matching algorithms this repo https://github.com/kdjones/fuzzystring.

Some thoughts from the follow-on conversation:
- Provide the user with control of the matching threshold
- Return ranked list of matches and matching score
- Uncertainty over which matching algorithm (or which combination) to use
- Consider the auto-creation of test data (target and set of inputs) based on specific use case
- Use of test data to help determine most useful algorithm

**Comment by pawelbaran** [comment]
Thank you @rolyhudson for the summary. One more word on the idea of user control of the matching threshold: the concept is to pass the responsibility for correctness of the matching process to the user - if one is confident enough to allow non-strict matches, change the default value from 1 to lower, with us doing our best to help, but not promising anything. This could be further highlighted by a warning on non-exact match etc.

Two BIM-specific user stories where the above could be helpful:
- pull from Revit => replacement of the pulled profiles with the BHoM library profiles using name matching with minimum threshold < 1
- push to Revit => name matching of family types with object properties
